### PR TITLE
Time Based Simulation Viewing

### PR DESCRIPTION
**Pull request by tatarize**
Please have a look at the bugfix branch, there's a beta of the updated simulation panel.
Allow a time-based view of simulation.
While the usual simulation jumps from one completed cutcode operation to the next, you can now choose a timed mode: the simulation jumps from one second to the next and displaying the progress so far:
 
simulation

There are some limitations still:
- intermediate cutcode operations are interpolated with a linear move from cutcode start to end 
- Raster operations are only shown as a growing rectangle
- Very fast units like balor might need a subsecond resolution
Image
Would like to get some feedback what to change, thanks

### CODE REVIEW

File: meerk40t/gui/simulation.py [hunk 1]
```diff
@@ -51,6 +55,8 @@ def __init__(self, *args, context=None, plan_name=None, auto_clear=True, **kwds)
             if isinstance(c, CutCode):
                 self.cutcode.extend(c)
         self.cutcode = CutCode(self.cutcode.flat())
+        self.statistics = self.cutcode.provide_statistics()
```
Comment (tatarize) [inline meerk40t/gui/simulation.py:58]: Does this update if the cutcode updates with the recalculate button?

File: meerk40t/gui/simulation.py [hunk 2]
```diff
@@ -383,14 +419,54 @@ def fit_scene_to_panel(self):
...
+            idx = 0
+        # print(
+        #     f"Cut-Mode={self._playback_cuts}, prog={progress}, idx={idx}, stats={len(self.statistics)}"
+        # )
+        return idx, residual
+
     def on_mouse_right_down(self, event=None):
+        def cut_before(event):
+            step, residual = self.progress_to_idx(self.progress)
+            self.context(f"plan{self.plan_name} sublist {step} -1\n")
```
Comment (tatarize) [inline meerk40t/gui/simulation.py:459]: Does this make sense in time mode? If we are not at a specific point in the cutcode where we can delete half the cutcode and be at this location, we might well try to do a `delete cuts before` partway through a raster or something which could be a bit strange. Though disabling this feature in that mode would suffice.

File: meerk40t/gui/simulation.py [hunk 3]
```diff
@@ -473,11 +567,10 @@ def _refresh_simulated_plan(self):
             if isinstance(c, CutCode):
                 self.cutcode.extend(c)
         self.cutcode = CutCode(self.cutcode.flat())
-        self.max = max(len(self.cutcode), 0) + 1
-        self.progress = self.max
-        self.slider_progress.SetMin(0)
-        self.slider_progress.SetMax(self.max)
-        self.slider_progress.SetValue(self.max)
+        self.statistics = self.cutcode.provide_statistics()
```
Comment (tatarize) [inline meerk40t/gui/simulation.py:570]: Ah it does

File: meerk40t/gui/simulation.py [hunk 4]
```diff
@@ -703,12 +824,72 @@ def __init__(self, scene, sim):
         self.matrix.post_cat(scene.context.device.device_to_scene_matrix())
 
     def process_draw(self, gc: wx.GraphicsContext):
-        if self.sim.progress > 1:
+        if self.sim.progress >= 0:
```
Comment (tatarize) [inline meerk40t/gui/simulation.py:827]: Could do:
```python
if self.sim.progress < 0:
    return
```
Then spare everything else in this block a tab.

**Comment by tatarize** [comment]
I would expect 100% speed and 1 second ticks to move at a speed of 1 second per tick.

**Comment by jpirnay** [comment]
Replaced by updated one

**Pull request closed without merging**